internal/schemarepo/loadschemas/plugins.go

```diff
@@ -190,6 +198,74 @@ func (cp *Plugins) ProviderSchema(addr addrs.Provider) (providers.ProviderSchema
 	return resp, nil
 }
 
+func (cp *Plugins) ResourceIdentitySchemas(addr addrs.Provider) (providers.ResourceIdentitySchemas, error) {
+	// Check the global schema cache first.
+	// This cache is only written by the provider client, and transparently
+	// used by GetProviderSchema, but we check it here because at this point we
+	// may be able to avoid spinning up the provider instance at all.
+	// We skip this if we have preloaded schemas because that suggests that
+	// our caller is not Terraform CLI and therefore it's probably inappropriate
+	// to assume that provider schemas are unique process-wide.
+	//
+	// FIXME: A global cache is inappropriate when Terraform Core is being
+	// used in a non-Terraform-CLI mode where we shouldn't assume that all
+	// calls share the same provider implementations.
+	schemas, ok := providers.ResourceIdentitySchemasCache.Get(addr)
+	if ok {
+		log.Printf("[TRACE] terraform.contextPlugins: Resource identity schemas for provider %q is in the global cache", addr)
+		return schemas, nil
+	}
+
+	if is, ok := cp.preloadedResourceIdentitySchemas[addr]; ok {
+		return is, nil
+	}
+
+	log.Printf("[TRACE] terraform.contextPlugins: Initializing provider %q to read its resource identity schemas", addr)
+	provider, err := cp.NewProviderInstance(addr)
+	if err != nil {
+		// A provider might have provided preloaded resource schemas but is missing the identity schemas.
+		// We will return an empty set of schemas in this case.
+		log.Printf("[TRACE] failed to instantiate provider %q to obtain resource identity schemas: %s", addr, err)
+		return providers.ResourceIdentitySchemas{}, nil
+	}
+	defer provider.Close()
+
+	resp := provider.GetResourceIdentitySchemas()
+	if resp.Diagnostics.HasErrors() {
+		for _, diag := range resp.Diagnostics {
+			if diag.Description().Summary == "Unsupported plugin method" {
```

Comment:
Optimally this should be handled at the grpc layer where we can detect a real unimplemented error type. I'd rather not pass stringified errors back up for matching and relying on the string being static.

Reply:
I've removed the explicit error check. It should be fine to return the error in that case, because callers of `ResourceIdentitySchemas` decide what to do if fetching identities errors.